File: advisor/pkg/k8s/seccomp_test.go
package k8s

import (
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

// SeccompProfile validation defends the contract the runtime expects:
// a non-empty DefaultAction, at least one architecture, and at least
// one syscall rule. A profile that fails any of these gates would be
// loaded by Docker/CRI-O but produce an unusable result (most likely
// blocking everything because no allow-list rule exists).

func TestNewProfileOptions_AcceptsWhitelistedActions(t *testing.T) {
	// All three values from the cobra help text must round-trip.
	for _, action := range []string{"SCMP_ACT_ERRNO", "SCMP_ACT_KILL", "SCMP_ACT_LOG"} {
		got, err := NewProfileOptions(&Config{OutputDir: "/tmp/x"}, action)
		assert.NoError(t, err, "action=%s", action)
		assert.Equal(t, action, got.DefaultAction, "DefaultAction must thread through")
		assert.Equal(t, "/tmp/x", got.OutputDir, "OutputDir must thread through from Config")
	}
}

func TestNewProfileOptions_RejectsUnknownAction(t *testing.T) {
	// Pre-fix the --default-action flag was a no-op — now it errors
	// loudly when the value isnt recognised. The error message must
	// list both the offending value and the accepted whitelist so the
	// operator doesnt have to consult docs.
	for _, bad := range []string{"", "SCMP_ACT_ALLOW", "SCMP_ACT_KILL_PROCESS", "errno", "scmp_act_kill"} {
		t.Run(bad, func(t *testing.T) {
			_, err := NewProfileOptions(&Config{}, bad)
			assert.Error(t, err)
			if err != nil {
				assert.Contains(t, err.Error(), bad, "error must echo the bad value")
				assert.Contains(t, err.Error(), "SCMP_ACT_ERRNO", "error must list accepted values")
			}
		})
	}
}

func TestNewProfileOptions_FallsBackToDefaultDirWhenConfigEmpty(t *testing.T) {
	// A nil/empty Config.OutputDir means the user didn't pass
	// --output-dir; preserve the historical default of
	// "seccomp-profiles" so previously-written shell scripts keep
	// working without any flag.
	got, err := NewProfileOptions(&Config{}, "SCMP_ACT_ERRNO")
	assert.NoError(t, err)
	assert.Equal(t, "seccomp-profiles", got.OutputDir)

	gotNil, err := NewProfileOptions(nil, "SCMP_ACT_ERRNO")
	assert.NoError(t, err)
	assert.Equal(t, "seccomp-profiles", gotNil.OutputDir, "nil Config must not panic + still default")
}

func TestValidateProfile_Valid(t *testing.T) {
	p := SeccompProfile{
		DefaultAction: "SCMP_ACT_ERRNO",
		Architectures: []string{"SCMP_ARCH_X86_64"},
		Syscalls: []Rule{
			{Names: []string{"read", "write"}, Action: "SCMP_ACT_ALLOW"},
		},
	}
	assert.NoError(t, ValidateProfile(p))
}

func TestValidateProfile_MissingDefaultAction(t *testing.T) {
	p := SeccompProfile{
		Architectures: []string{"SCMP_ARCH_X86_64"},
		Syscalls:      []Rule{{Names: []string{"read"}, Action: "SCMP_ACT_ALLOW"}},
	}
	err := ValidateProfile(p)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "default action")
}

func TestValidateProfile_NoArchitectures(t *testing.T) {
	p := SeccompProfile{
		DefaultAction: "SCMP_ACT_ERRNO",
		Architectures: nil,
		Syscalls:      []Rule{{Names: []string{"read"}, Action: "SCMP_ACT_ALLOW"}},
	}
	err := ValidateProfile(p)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "architecture")
}

func TestValidateProfile_NoSyscalls(t *testing.T) {
	// A profile with no rules and a default-deny action would block
	// every syscall. Could be intentional but we treat empty as a
	// likely operator mistake.
	p := SeccompProfile{
		DefaultAction: "SCMP_ACT_ERRNO",
		Architectures: []string{"SCMP_ARCH_X86_64"},
		Syscalls:      []Rule{},
	}
	err := ValidateProfile(p)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "syscall")
}

func TestValidateProfile_EmptyArchitectureSlice(t *testing.T) {
	// Distinguish nil vs empty — both must be rejected.
	p := SeccompProfile{
		DefaultAction: "SCMP_ACT_ERRNO",
		Architectures: []string{},
		Syscalls:      []Rule{{Names: []string{"read"}, Action: "SCMP_ACT_ALLOW"}},
	}
	err := ValidateProfile(p)
	assert.Error(t, err)
}

// BuildSeccompProfile must map the arch strings the controller actually
// records — Rust's std::env::consts::ARCH: "x86_64" / "aarch64"
// (controller/src/syscall.rs). A regression here reintroduces the bug where
// the map keyed on the never-written "ARM64", so ARM nodes got a profile with
// "architectures": null that ValidateProfile rejects.
func TestBuildSeccompProfile_MapsControllerArchStrings(t *testing.T) {
	cases := map[string]string{
		"x86_64":  "SCMP_ARCH_X86_64",
		"aarch64": "SCMP_ARCH_ARM64",
	}
	for arch, want := range cases {
		profile := BuildSeccompProfile([]string{"read", "write"}, arch, "SCMP_ACT_ERRNO")
		assert.Equal(t, []string{want}, profile.Architectures, "arch %q should map to %q", arch, want)
		assert.NoError(t, ValidateProfile(profile), "arch %q should produce a valid profile", arch)
	}
}

// MergeSyscalls deduplicates the union of multiple syscall lists.
// Order of the output is map-iteration-order (non-deterministic), so
// tests compare against sorted slices.

func TestMergeSyscalls_Empty(t *testing.T) {
	assert.Empty(t, MergeSyscalls())
}

func TestMergeSyscalls_SingleListPassthrough(t *testing.T) {
	got := MergeSyscalls([]string{"read", "write", "open"})
	sort.Strings(got)
	assert.Equal(t, []string{"open", "read", "write"}, got)
}

func TestMergeSyscalls_DeduplicatesAcrossLists(t *testing.T) {
	got := MergeSyscalls(
		[]string{"read", "write"},
		[]string{"write", "open"},
		[]string{"read", "close"},
	)
	sort.Strings(got)
	assert.Equal(t, []string{"close", "open", "read", "write"}, got)
}

func TestMergeSyscalls_DeduplicatesWithinSingleList(t *testing.T) {
	got := MergeSyscalls([]string{"read", "read", "write", "write"})
	sort.Strings(got)
	assert.Equal(t, []string{"read", "write"}, got)
}

func TestMergeSyscalls_HandlesEmptyInputs(t *testing.T) {
	got := MergeSyscalls(nil, []string{}, []string{"read"}, nil)
	sort.Strings(got)
	assert.Equal(t, []string{"read"}, got)
}

// The rules-array check counts one, not zero, for the profile that matters:
// a single rule with an empty Names list. Under a denying DefaultAction that
// denies every syscall, so the container never starts — the filter is
// installed before execve. Parity with the frontend and llm-bridge
// validators, which had the identical gap.
func TestValidateProfileRejectsDenyingDefaultThatAllowsNothing(t *testing.T) {
	err := ValidateProfile(SeccompProfile{
		DefaultAction: "SCMP_ACT_ERRNO",
		Architectures: []string{"SCMP_ARCH_X86_64"},
		Syscalls:      []Rule{{Names: []string{}, Action: "SCMP_ACT_ALLOW"}},
	})
	assert.ErrorContains(t, err, "no rule allows any syscall")
}

func TestValidateProfileAcceptsDenyingDefaultWithAnAllowedSyscall(t *testing.T) {
	assert.NoError(t, ValidateProfile(SeccompProfile{
		DefaultAction: "SCMP_ACT_ERRNO",
		Architectures: []string{"SCMP_ARCH_X86_64"},
		Syscalls:      []Rule{{Names: []string{"read"}, Action: "SCMP_ACT_ALLOW"}},
	}))
}

// A permissive default with an empty rule is a no-op rather than a trap, so
// the new check must not fire on audit-only profiles.
func TestValidateProfileAcceptsAuditOnlyProfile(t *testing.T) {
	assert.NoError(t, ValidateProfile(SeccompProfile{
		DefaultAction: "SCMP_ACT_LOG",
		Architectures: []string{"SCMP_ARCH_X86_64"},
		Syscalls:      []Rule{{Names: []string{}, Action: "SCMP_ACT_ALLOW"}},
	}))
}

File: llm-bridge/src/tools/generators/seccomp.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { seccompFromBrokerSyscalls, validateSeccompProfile } from "./seccomp.js";

// Regression guard (fix): the assistant's in-process seccomp generator must
// reproduce the advisor serve handler's build-then-validate behavior. An
// unrecognized CPU arch produces `architectures: []`, which the advisor's
// k8s.ValidateProfile rejects — porting the tools in-process originally dropped
// that guard, so an exotic-arch pod would get a silently-broken profile.

test("seccompFromBrokerSyscalls: unknown arch throws instead of a silent broken profile", () => {
  assert.throws(
    () => seccompFromBrokerSyscalls({ syscalls: "read,write", arch: "ppc64le" }),
    /unrecognized architecture/,
  );
  // Missing/empty arch is the same failure mode.
  assert.throws(() => seccompFromBrokerSyscalls({ syscalls: "read", arch: "" }), /unrecognized architecture/);
  assert.throws(() => seccompFromBrokerSyscalls({ syscalls: "read" }), /unrecognized architecture/);
});

test("seccompFromBrokerSyscalls: known arch returns a valid, sorted profile", () => {
  const x86 = seccompFromBrokerSyscalls({ syscalls: "write,read,openat", arch: "x86_64" });
  assert.deepEqual(x86.architectures, ["SCMP_ARCH_X86_64"]);
  assert.equal(x86.defaultAction, "SCMP_ACT_ERRNO");
  assert.deepEqual(x86.syscalls[0].names, ["openat", "read", "write"]);

  const arm = seccompFromBrokerSyscalls({ syscalls: "read", arch: "aarch64" });
  assert.deepEqual(arm.architectures, ["SCMP_ARCH_ARM64"]);
});

test("validateSeccompProfile: mirrors advisor ValidateProfile conditions", () => {
  assert.throws(
    () => validateSeccompProfile({ defaultAction: "", architectures: ["SCMP_ARCH_X86_64"], syscalls: [{ names: ["read"], action: "SCMP_ACT_ALLOW" }] }),
    /default action is required/,
  );
  assert.throws(
    () => validateSeccompProfile({ defaultAction: "SCMP_ACT_ERRNO", architectures: [], syscalls: [{ names: ["read"], action: "SCMP_ACT_ALLOW" }] }),
    /unrecognized architecture/,
  );
  assert.throws(
    () => validateSeccompProfile({ defaultAction: "SCMP_ACT_ERRNO", architectures: ["SCMP_ARCH_X86_64"], syscalls: [] }),
    /at least one syscall rule/,
  );
});

// The rules-array check counts one, not zero, for the profile that matters:
// a single rule with an empty `names` list. Under a denying defaultAction
// that permits nothing at all, and the container never starts, because the
// filter is installed before execve. Reachable straight from a broker
// response of `{ syscalls: "", arch: "x86_64" }`, and an MCP tool's output
// may be applied without a person reading it.
test("rejects a denying default that allows no syscall by name", () => {
  assert.throws(
    () => validateSeccompProfile({
      defaultAction: "SCMP_ACT_ERRNO",
      architectures: ["SCMP_ARCH_X86_64"],
      syscalls: [{ names: [], action: "SCMP_ACT_ALLOW" }],
    }),
    /no rule allows any syscall/,
  );
});

test("rejects it via the broker-response path too", () => {
  assert.throws(
    () => seccompFromBrokerSyscalls({ syscalls: "", arch: "x86_64" }),
    /no rule allows any syscall/,
  );
});

test("accepts a denying default once one rule allows something", () => {
  assert.doesNotThrow(() =>
    seccompFromBrokerSyscalls({ syscalls: "read,write", arch: "x86_64" }),
  );
});

// A permissive default with an empty rule is a no-op rather than a trap, so
// the check must not fire: audit-only profiles legitimately allow nothing by
// name and still let the workload run.
test("accepts a permissive default with an empty rule", () => {
  assert.doesNotThrow(() =>
    validateSeccompProfile({
      defaultAction: "SCMP_ACT_LOG",
      architectures: ["SCMP_ARCH_X86_64"],
      syscalls: [{ names: [], action: "SCMP_ACT_ALLOW" }],
    }),
  );
});
